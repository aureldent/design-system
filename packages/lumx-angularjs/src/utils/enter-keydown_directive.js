import { ENTER_KEY_CODE } from '@lumx/core/js/constants';

/////////////////////////////

function EnterKeydownDirective() {
    'ngInject';

    function link(scope, el, attrs) {
        el.on('keydown', (evt) => {
            if (evt.which === ENTER_KEY_CODE) {
                scope.$apply(() => {
                    scope.$eval(attrs.lxEnterKeydown, { $event: evt });
                });

                evt.preventDefault();
            }
        });
    }

    return {
        link,
    };
}

/////////////////////////////

angular.module('lumx.utils.enter-keydown').directive('lxEnterKeydown', EnterKeydownDirective);

/////////////////////////////

export { EnterKeydownDirective };
