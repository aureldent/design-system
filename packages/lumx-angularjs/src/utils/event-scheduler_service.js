function EventSchedulerService($document, LxUtilsService) {
    'ngInject';

    const service = this;

    /////////////////////////////
    //                         //
    //    Private attributes   //
    //                         //
    /////////////////////////////

    /**
     * The handlers.
     *
     * @type {Object}
     */
    const _handlers = {};

    /**
     * The schedule.
     *
     * @type {Object}
     */
    const _schedule = {};

    /////////////////////////////
    //                         //
    //    Private functions    //
    //                         //
    /////////////////////////////

    /**
     * Handle en event.
     *
     * @param {Event} evt The event.
     */
    function _handle(evt) {
        let scheduler;

        for (const [key] of Object.entries(_schedule)) {
            if (key.includes(evt.type)) {
                scheduler = _schedule[key];
            }
        }

        if (angular.isDefined(scheduler)) {
            for (let i = 0, len = scheduler.length; i < len; i++) {
                const handler = scheduler[i];

                if (angular.isDefined(handler) && angular.isDefined(handler.cb) && angular.isFunction(handler.cb)) {
                    handler.cb(evt);

                    if (evt.isPropagationStopped()) {
                        break;
                    }
                }
            }
        }
    }

    /////////////////////////////
    //                         //
    //     Public functions    //
    //                         //
    /////////////////////////////

    /**
     * Register en event.
     *
     * @param  {string}   eventName The event name.
     * @param  {Function} cb        The event callback.
     * @return {string}   The event id.
     */
    function register(eventName, cb) {
        const handler = {
            cb,
            eventName,
        };

        const id = LxUtilsService.generateUUID();
        _handlers[id] = handler;

        if (angular.isUndefined(_schedule[eventName])) {
            _schedule[eventName] = [];

            $document.on(eventName, _handle);
        }

        _schedule[eventName].unshift(_handlers[id]);

        return id;
    }

    /**
     * Unregister en event.
     *
     * @param  {string} id The event id.
     * @return {found}  Whether the event has been found or not.
     */
    function unregister(id) {
        let found = false;
        const handler = _handlers[id];

        if (angular.isDefined(handler) && angular.isDefined(_schedule[handler.eventName])) {
            const index = _schedule[handler.eventName].indexOf(handler);

            if (angular.isDefined(index) && index > -1) {
                _schedule[handler.eventName].splice(index, 1);

                delete _handlers[id];
                found = true;
            }

            if (_schedule[handler.eventName].length === 0) {
                delete _schedule[handler.eventName];

                $document.off(handler.eventName, _handle);
            }
        }

        return found;
    }

    /////////////////////////////

    service.register = register;
    service.unregister = unregister;
}

/////////////////////////////

angular.module('lumx.utils.event-scheduler').service('LxEventSchedulerService', EventSchedulerService);

/////////////////////////////

export { EventSchedulerService };
