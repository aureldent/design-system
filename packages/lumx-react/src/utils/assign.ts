export const assign = (target: any, ...sources: any[]) => {
    sources.forEach((source) => Object.keys(source).forEach((key) => (target[key] = source[key])));
    return target;
};
