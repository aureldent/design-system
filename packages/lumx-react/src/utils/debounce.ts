export function debounce(func: (...args: any) => any, debounceTime: number) {
    let timeout: any;
    return function debouncedFunction(...args: any[]) {
        clearTimeout(timeout);
        timeout = setTimeout(() => func(...args), debounceTime);
    };
}
