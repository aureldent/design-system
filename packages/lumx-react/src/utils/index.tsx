export { getBasicClass, handleBasicClasses, detectSwipe, onEnterPressed, onEscapePressed } from '@lumx/core/utils';

export * from './calculatePopoverPlacement';

export * from './isInViewPort';

export * from './validation';

export * from './partitionMulti';

export { getRootClassName } from './getRootClassName';

export * from './type';
