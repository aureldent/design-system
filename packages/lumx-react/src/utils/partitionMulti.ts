import { Predicate } from './type';
import reduce from 'lodash/reduce';
import concat from 'lodash/concat';
import dropRight from 'lodash/dropRight';
import partition from 'lodash/partition';
import last from 'lodash/last';

/**
 * Similar to lodash `partition` function but working with multiple predicates.
 *
 * @example
 * const isString = (s) => typeof s === 'string'
 * const isNumber = (s) => typeof s === 'number'
 * const [strings, numbers, others] = partitionMulti(['a', 1, 'b', false], [isString, isNumber])
 * //=> [['a', 'b'], [1], [false]]
 *
 * @param  elements array of elements
 * @param  predicates array of predicates to apply on elements
 * @return partitioned elements by the given predicates
 */
function partitionMulti<T>(elements: T[], predicates: Array<Predicate<T>>): T[][] {
    return reduce(
        predicates,
        (partitioned: T[][], predicate: Predicate<T>) =>
            concat(dropRight(partitioned), partition(last(partitioned), predicate)),
        [elements],
    );
}

export { partitionMulti };
