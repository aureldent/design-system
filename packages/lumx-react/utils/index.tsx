export * from './calculatePopoverPlacement';

export * from './isInViewPort';
