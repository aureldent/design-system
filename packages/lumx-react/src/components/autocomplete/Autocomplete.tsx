import React, { ReactElement, RefObject, useRef } from 'react';

import classNames from 'classnames';

import { Dropdown, Offset, Placement, TextField, TextFieldType, Theme } from '@lumx/react';

import { COMPONENT_PREFIX } from '@lumx/react/constants';
import { handleBasicClasses } from '@lumx/react/utils';
import { IGenericProps, getRootClassName } from '@lumx/react/utils';

import { useFocusOnClose } from '@lumx/react/hooks/useFocusOnClose';

/////////////////////////////

/**
 * Defines the props of the component.
 */
interface IAutocompleteProps extends IGenericProps {
    /** A ref that will be passed to the input or textarea element. */
    inputRef?: RefObject<HTMLInputElement>;

    /**
     * Vertical and/or horizontal offsets that will be applied to the Dropdown position.
     * @see {@link DropdownProps#offset}
     */
    offset?: Offset;

    /**
     * The preferred Dropdown location against the anchor element.
     * @see {@link DropdownProps#placement}
     */
    placement?: Placement;

    /**
     * Whether the dropdown should fit to the anchor width
     * @see {@link DropdownProps#hasError}
     */
    fitToAnchorWidth?: boolean;

    /**
     * Whether the text field is displayed with error style or not.
     * @see {@link TextFieldProps#hasError}
     */
    hasError?: boolean;

    /**
     * Whether the text field displays a clear button or not.
     * @see {@link TextFieldProps#hasError}
     */
    isClearable?: boolean;

    /**
     * Text field helper message.
     * @see {@link TextFieldProps#helper}
     */
    helper?: string;

    /**
     * Text field icon (SVG path)
     * @see {@link TextFieldProps#icon}
     */
    icon?: string;

    /**
     * Whether the text field is disabled or not.
     * @see {@link TextFieldProps#isDisabled}
     */
    isDisabled?: boolean;

    /**
     * Whether the text field is displayed with valid style or not.
     * @see {@link TextFieldProps#isValid}
     */
    isValid?: boolean;

    /**
     * Text field label displayed in a label tag.
     * @see {@link TextFieldProps#label}
     */
    label?: string;

    /**
     * Text field placeholder message.
     * @see {@link TextFieldProps#placeholder}
     */
    placeholder?: string;

    /** Theme. */
    theme?: Theme;

    /**
     * Children of the Autocomplete. This should be a list of the different
     * suggestions that
     */
    children: React.ReactNode;

    /**
     * List of chips to be displayed before the text field input.
     */
    chips?: React.ReactNode;

    /**
     * Text field value.
     * @see {@link TextFieldProps#onChange}
     */
    value: string;

    /**
     * Whether the suggestions from the autocomplete should be displayed or not.
     * Useful to control when the suggestions are displayed from outside the component
     */
    isOpen: boolean;

    /**
     * Whether a click anywhere out of the Autocomplete would close it
     * @see {@link DropdownProps#closeOnClick}
     */
    closeOnClick?: boolean;

    /**
     * Whether an escape key press would close the Autocomplete.
     * @see {@link DropdownProps#closeOnEscape}
     */
    closeOnEscape?: boolean;

    /**
     * The function to be called when the user clicks away or Escape is pressed
     * @see {@link DropdownProps#onClose}
     */
    onClose?: VoidFunction;

    /**
     * The callback function called when the bottom of the dropdown is reached.
     * @see {@link DropdownProps#onInfinite}
     */
    onInfinite?: VoidFunction;

    /**
     * Text field value change handler.
     * @see {@link TextFieldProps#onChange}
     */
    onChange(value: string): void;

    /**
     * Text field focus change handler.
     * @see {@link TextFieldProps#onFocus}
     */
    onFocus?(event: React.FocusEvent): void;

    /**
     * Text field blur change handler.
     * @see {@link TextFieldProps#onBlur}
     */
    onBlur?(event: React.FocusEvent): void;
}
type AutocompleteProps = IAutocompleteProps;

/////////////////////////////

/////////////////////////////
//                         //
//    Public attributes    //
//                         //
/////////////////////////////

/**
 * The display name of the component.
 */
const COMPONENT_NAME = `${COMPONENT_PREFIX}Autocomplete`;

/**
 * The default class name and classes prefix for this component.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * The default value of props.
 */
const DEFAULT_PROPS: Partial<AutocompleteProps> = {
    closeOnClick: true,
    closeOnEscape: true,
    isOpen: undefined,
};

/////////////////////////////

/**
 * This component allows to make the connection between a Text Field and a Dropdown,
 * displaying a list of suggestions from the text entered on the text field.
 *
 * @return The component.
 */
const Autocomplete: React.FC<AutocompleteProps> = (props: AutocompleteProps): ReactElement => {
    const {
        className,
        children,
        chips,
        value,
        onBlur,
        onChange,
        onFocus,
        isOpen,
        closeOnClick,
        closeOnEscape,
        hasError,
        helper,
        icon,
        isDisabled,
        isClearable,
        isValid,
        label,
        placeholder,
        theme,
        onClose,
        offset,
        placement,
        inputRef = useRef(null),
        fitToAnchorWidth,
        onInfiniteScroll,
        ...forwardedProps
    } = props;

    const textFieldRef = useRef(null);
    useFocusOnClose(inputRef.current, isOpen);

    return (
        <div
            className={classNames(
                className,
                handleBasicClasses({
                    prefix: CLASSNAME,
                }),
            )}
            {...forwardedProps}
        >
            <TextField
                value={value}
                onChange={onChange}
                chips={chips}
                textFieldRef={textFieldRef}
                inputRef={inputRef}
                isClearable={isClearable}
                onBlur={onBlur}
                onFocus={onFocus}
                hasError={hasError}
                helper={helper}
                icon={icon}
                isDisabled={isDisabled}
                isValid={isValid}
                label={label}
                placeholder={placeholder}
                theme={theme}
                type={TextFieldType.input}
            />
            <Dropdown
                anchorRef={textFieldRef as React.RefObject<HTMLElement>}
                showDropdown={isOpen}
                closeOnClick={closeOnClick}
                closeOnEscape={closeOnEscape}
                onClose={onClose}
                offset={offset}
                placement={placement}
                fitToAnchorWidth={fitToAnchorWidth}
                onInfiniteScroll={onInfiniteScroll}
                theme={theme}
            >
                {children}
            </Dropdown>
        </div>
    );
};
Autocomplete.displayName = COMPONENT_NAME;

/////////////////////////////

export { CLASSNAME, DEFAULT_PROPS, Autocomplete, AutocompleteProps };
