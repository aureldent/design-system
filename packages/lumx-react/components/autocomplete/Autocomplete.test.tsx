import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';

import { List, ListItem, Size } from 'LumX';
import { ICommonSetup, Wrapper, commonTestsSuite } from 'LumX/core/testing/utils.test';

import { Autocomplete, AutocompleteProps, CLASSNAME } from './Autocomplete';

import { CITIES as suggestions } from './__mockData__';

/////////////////////////////

/**
 * Define the overriding properties expected by the `setup` function.
 */
type ISetupProps = Partial<AutocompleteProps>;

/**
 * Defines what the `setup` function will return.
 */
interface ISetup extends ICommonSetup {
    props: ISetupProps;

    /**
     * The <div> element that holds the autocomplete content.
     */
    wrapper: Wrapper;

    /**
     * Text Field component.
     */
    textField: Wrapper;

    /**
     * Dropdown component.
     */
    dropdown: Wrapper;
}

interface ISuggestion {
    id: number;
    text: string;
}

/////////////////////////////

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param  props  The props to use to override the default props of the component.
 * @param  [shallowRendering=true] Indicates if we want to do a shallow or a full rendering.
 * @return An object with the props, the component wrapper and some shortcut to some element inside of the component.
 */
const setup = (props: ISetupProps = {}, shallowRendering: boolean = true): ISetup => {
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;

    // @ts-ignore
    const wrapper: Wrapper = renderer(<Autocomplete {...props} />);

    const textField: Wrapper = wrapper.find('TextField');
    const dropdown: Wrapper = wrapper.find('Dropdown');

    return {
        dropdown,
        props,
        textField,
        wrapper,
    };
};

describe(`<${Autocomplete.displayName}>`, (): void => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', (): void => {
        // Here is an example of a basic rendering check, with snapshot.

        it('should render correctly', (): void => {
            const { wrapper, textField, dropdown } = setup({
                children: (
                    <List>
                        {suggestions.map((suggestion: ISuggestion) => (
                            <ListItem size={Size.tiny} isClickable key={suggestion.id}>
                                <div>{suggestion.text}</div>
                            </ListItem>
                        ))}
                    </List>
                ),
                isOpen: true,
                onChange: jest.fn(),
                value: '',
            });
            expect(wrapper).toMatchSnapshot();

            expect(wrapper).toExist();

            expect(wrapper).toHaveClassName(CLASSNAME);

            expect(textField).toHaveLength(1);
            expect(dropdown).toHaveLength(1);
        });
    });

    /////////////////////////////

    // 2. Test defaultProps value and important props custom values.
    describe('Props', (): void => {
        it('should use default props', (): void => {
            const { wrapper }: ISetup = setup();

            expect(wrapper).toHaveClassName(CLASSNAME);
        });

        it('should render correctly when the dropdown is closed', (): void => {
            const { wrapper } = setup({
                children: (
                    <List>
                        {suggestions.map((suggestion: ISuggestion) => (
                            <ListItem size={Size.tiny} isClickable key={suggestion.id}>
                                <div>{suggestion.text}</div>
                            </ListItem>
                        ))}
                    </List>
                ),
                isOpen: false,
                onChange: jest.fn(),
                value: '',
            });
            expect(wrapper).toMatchSnapshot();

            expect(wrapper).toExist();

            expect(wrapper).toHaveClassName(CLASSNAME);
        });
    });

    /////////////////////////////

    // 3. Test events.
    describe('Events', (): void => {
        it('should trigger the onChange callback when there is a change on the Text Field', (): void => {
            const onChange = jest.fn();
            const { textField } = setup({
                children: (
                    <List>
                        {suggestions.map((suggestion: ISuggestion) => (
                            <ListItem size={Size.tiny} isClickable key={suggestion.id}>
                                <div>{suggestion.text}</div>
                            </ListItem>
                        ))}
                    </List>
                ),
                isOpen: false,
                onChange,
                value: '',
            });

            textField.simulate('change');

            expect(onChange).toHaveBeenCalled();
        });

        it('should trigger the onFocus callback when the text field is focused on', (): void => {
            const onFocus = jest.fn();
            const { textField } = setup({
                children: (
                    <List>
                        {suggestions.map((suggestion: ISuggestion) => (
                            <ListItem size={Size.tiny} isClickable key={suggestion.id}>
                                <div>{suggestion.text}</div>
                            </ListItem>
                        ))}
                    </List>
                ),
                isOpen: false,
                onFocus,
                value: '',
            });

            textField.simulate('focus');

            expect(onFocus).toHaveBeenCalled();
        });

        it('should trigger the onBlur callback when the Text Field loses focus', (): void => {
            const onBlur = jest.fn();
            const { textField } = setup({
                children: (
                    <List>
                        {suggestions.map((suggestion: ISuggestion) => (
                            <ListItem size={Size.tiny} isClickable key={suggestion.id}>
                                <div>{suggestion.text}</div>
                            </ListItem>
                        ))}
                    </List>
                ),
                isOpen: false,
                onBlur,
                value: '',
            });

            textField.simulate('blur');

            expect(onBlur).toHaveBeenCalled();
        });
    });
    /////////////////////////////

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', (): void => {
        // Nothing to do here.
    });

    /////////////////////////////

    // 5. Test state.
    describe('State', (): void => {
        // Nothing to do here.
    });

    /////////////////////////////

    // Common tests suite.
    commonTestsSuite(setup, { className: 'wrapper', prop: 'wrapper' }, { className: CLASSNAME });
});
