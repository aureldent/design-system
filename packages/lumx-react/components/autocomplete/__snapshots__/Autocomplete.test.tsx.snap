// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Autocomplete> Props should render correctly when the dropdown is closed 1`] = `
<div
  className="lumx-autocomplete"
>
  <TextField
    inputRef={
      Object {
        "current": null,
      }
    }
    onChange={[MockFunction]}
    textFieldRef={
      Object {
        "current": null,
      }
    }
    type="input"
    value=""
  />
  <Dropdown
    anchorRef={
      Object {
        "current": null,
      }
    }
    showDropdown={false}
  >
    <List>
      <ListItem
        isClickable={true}
        key="losangeles"
        size="tiny"
      >
        <div>
          Los Angeles
        </div>
      </ListItem>
      <ListItem
        isClickable={true}
        key="sanfrancisco"
        size="tiny"
      >
        <div>
          San Francisco
        </div>
      </ListItem>
      <ListItem
        isClickable={true}
        key="paris"
        size="tiny"
      >
        <div>
          Paris
        </div>
      </ListItem>
      <ListItem
        isClickable={true}
        key="montpellier"
        size="tiny"
      >
        <div>
          Montpellier
        </div>
      </ListItem>
      <ListItem
        isClickable={true}
        key="bordeaux"
        size="tiny"
      >
        <div>
          Bordeaux
        </div>
      </ListItem>
      <ListItem
        isClickable={true}
        key="toulouse"
        size="tiny"
      >
        <div>
          Toulouse
        </div>
      </ListItem>
      <ListItem
        isClickable={true}
        key="lyon"
        size="tiny"
      >
        <div>
          Lyon
        </div>
      </ListItem>
      <ListItem
        isClickable={true}
        key="montevideo"
        size="tiny"
      >
        <div>
          Montevideo
        </div>
      </ListItem>
    </List>
  </Dropdown>
</div>
`;

exports[`<Autocomplete> Snapshots and structure should render correctly 1`] = `
<div
  className="lumx-autocomplete"
>
  <TextField
    inputRef={
      Object {
        "current": null,
      }
    }
    onChange={[MockFunction]}
    textFieldRef={
      Object {
        "current": null,
      }
    }
    type="input"
    value=""
  />
  <Dropdown
    anchorRef={
      Object {
        "current": null,
      }
    }
    showDropdown={true}
  >
    <List>
      <ListItem
        isClickable={true}
        key="losangeles"
        size="tiny"
      >
        <div>
          Los Angeles
        </div>
      </ListItem>
      <ListItem
        isClickable={true}
        key="sanfrancisco"
        size="tiny"
      >
        <div>
          San Francisco
        </div>
      </ListItem>
      <ListItem
        isClickable={true}
        key="paris"
        size="tiny"
      >
        <div>
          Paris
        </div>
      </ListItem>
      <ListItem
        isClickable={true}
        key="montpellier"
        size="tiny"
      >
        <div>
          Montpellier
        </div>
      </ListItem>
      <ListItem
        isClickable={true}
        key="bordeaux"
        size="tiny"
      >
        <div>
          Bordeaux
        </div>
      </ListItem>
      <ListItem
        isClickable={true}
        key="toulouse"
        size="tiny"
      >
        <div>
          Toulouse
        </div>
      </ListItem>
      <ListItem
        isClickable={true}
        key="lyon"
        size="tiny"
      >
        <div>
          Lyon
        </div>
      </ListItem>
      <ListItem
        isClickable={true}
        key="montevideo"
        size="tiny"
      >
        <div>
          Montevideo
        </div>
      </ListItem>
    </List>
  </Dropdown>
</div>
`;
